# **Water Quality - Data Visualisation**

## Objectives

* Write your notebook objective here, for example, "Fetch data from Kaggle and save as raw data", or "engineer features for modelling"

## Inputs

* Write down which data or information you need to run the notebook 

## Outputs

* Write here which files, code or artefacts you generate by the end of the notebook 

## Additional Comments

* If you have any additional comments that don't fit in the previous bullets, please state them here. 



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [2]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\Lailah\\vscode-project\\DA2_Assessment\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [3]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [4]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\Lailah\\vscode-project\\DA2_Assessment'

# Section 1

Loading Clean Data

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [6]:
df = pd.read_csv("data/Cleaned_Data/Cleaned_Date.csv")
df.head(8)

,Unnamed: 0,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
0,0,7.080795,204.890455,20791.318981,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135,0
1,1,3.889107,129.422921,18630.057858,6.635246,333.775777,592.885359,15.180013,56.329076,4.500656,0
2,2,8.099124,224.236259,19909.541732,9.275884,333.775777,418.606213,16.868637,66.420093,3.055934,0
3,3,8.316766,214.373394,22018.417441,8.059332,356.886136,363.266516,18.436524,100.341674,4.628771,0
4,4,9.092223,181.101509,17978.986339,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075,0
5,5,5.584087,188.313324,28748.687739,7.544869,326.678363,280.467916,8.399735,54.917862,2.559708,0
6,6,10.223862,248.071735,28749.716544,7.513408,393.663396,283.651634,13.789695,84.603556,2.672989,0
7,7,8.635849,203.361523,13672.091764,4.563009,303.309771,474.607645,12.363817,62.798309,4.401425,0


In [7]:
df.tail(8)

,Unnamed: 0,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
3268,3268,6.702547,207.321086,17246.920347,7.708117,304.510230,329.266002,16.217303,28.878601,3.442983,1
3269,3269,10.258615,117.125160,37188.826022,9.263166,267.157960,439.893618,16.172755,41.558501,4.369264,1
3270,3270,6.069616,186.659040,26138.780191,7.747547,345.700257,415.886955,12.067620,60.419921,3.669712,1
3271,3271,4.668102,193.681735,44831.869873,7.166639,359.948574,526.424171,13.894419,66.687695,4.435821,1
3272,3272,7.808856,193.553212,17329.802160,8.061362,333.775777,392.449580,19.903225,66.396293,2.798243,1
3273,3273,9.419510,175.762646,33155.578218,7.350233,333.775777,432.044783,11.039070,69.845400,3.298875,1
3274,3274,5.126763,230.603758,11983.869376,6.303357,333.775777,402.883113,11.168946,77.488213,4.708658,1
3275,3275,7.874671,195.102299,17404.177061,7.509306,333.775777,327.459760,16.140368,78.698446,2.309149,1


In [8]:
types = df.dtypes
print(types)

Unnamed: 0           int64
ph                 float64
Hardness           float64
Solids             float64
Chloramines        float64
Sulfate            float64
Conductivity       float64
Organic_carbon     float64
Trihalomethanes    float64
Turbidity          float64
Potability           int64
dtype: object


In [9]:
df.shape

(3276, 11)

---

# Section 2

Section 2


In [44]:
#df_corr=df[['bmi','age','avg_glucose_level','heart_disease','hypertension']].corr(method='pearson')
#df_corr

In [ ]:
# sns.set_theme(style="whitegrid")
# sns.heatmap(df_corr, annot=True, cmap='coolwarm', linewidths=0.5, annot_kws={"size":8})
# plt.show()

In [38]:
# Convert 0/1 into readable labels BEFORE plotting
#df["Potability"] = df["Potability"].map({
#     0: "Not potable",
#     1: "Potable"
# })

# # Now calculate proportions
# Potability_count = df.value_counts('Potability', normalize=True)

# # Plot the pie chart
# plt.pie(
#     x=Potability_count,
#     labels=Potability_count.index,
#     autopct='%1.1f%%'
# )
# plt.show()


In [ ]:
#fig=sns.pairplot(data=df[['RestingBP', 'Cholesterol', 'Age', 'MaxHR', 'HeartDisease']], hue='HeartDisease', plot_kws={'alpha': 0.5});
#for i, j in zip(*np.triu_indices_from(fig.axes, 1)):
    #fig.axes[i, j].set_visible(False)
#plt.show()

In [ ]:
# sns.set_theme(style="white")
# mask = np.triu(np.ones_like(df_corr, dtype=bool))
# sns.heatmap(df_corr, mask=mask, annot=True, cmap='coolwarm', linewidths=0.5, annot_kws={"size": 8})
# plt.ylim(df_corr.shape[0], 0)
# plt.show()

Bivariate Analysis: Comparing the Impact of different features on water potability 

In [40]:

# fig, axes = plt.subplots(nrows=8, ncols=2, figsize=(15, 20))

# sns.boxplot(data=df, x='ph', y='Potability', ax=axes[0,0])
# #sns.swarmplot(data=df, x=x, y=y, ax=axes[1], dodge=True, size=3)
# sns.stripplot(data=df, x='ph', y='Potability', ax=axes[0,1])

# sns.boxplot(data=df, x='Hardness', y='Potability', ax=axes[1,0])
# sns.stripplot(data=df, x='Hardness', y='Potability', ax=axes[1,1])

# sns.boxplot(data=df, x='Solids', y='Potability', ax=axes[2,0])
# sns.stripplot(data=df, x='Solids', y='Potability', ax=axes[2,1])

# sns.boxplot(data=df, x='Chloramines', y='Potability', ax=axes[3,0])
# sns.stripplot(data=df, x='Chloramines', y='Potability', ax=axes[3,1])

# sns.boxplot(data=df, x='Sulfate', y='Potability', ax=axes[4,0])
# sns.stripplot(data=df, x='Sulfate', y='Potability', ax=axes[4,1])

# sns.boxplot(data=df, x='Conductivity', y='Potability', ax=axes[5,0])
# sns.stripplot(data=df, x='Conductivity', y='Potability', ax=axes[5,1])

# sns.boxplot(data=df, x='Organic_carbon', y='Potability', ax=axes[6,0])
# sns.stripplot(data=df, x='Organic_carbon', y='Potability', ax=axes[6,1])

# sns.boxplot(data=df, x='Trihalomethanes', y='Potability', ax=axes[7,0])
# sns.stripplot(data=df, x='Trihalomethanes', y='Potability', ax=axes[7,1])


# plt.tight_layout(pad=3.0)
# plt.show()



Python had issues running the swarmplot with my data; The issue: When there are too many data points or the markers are too large, seaborn cannot find non-overlapping positions for all points. The percentages indicate how many points couldn't be placed:

16.9% in the first plot
13.0% in the second plot
33.4% in the third plot (most severe)
21.0% in the fourth plot

so coppilot suggested using a striplot which worked out much better

Most water‑quality features show substantial overlap between potable and non‑potable samples. No single variable provides clear separation, indicating that potability depends on a combination of factors rather than any one parameter. This supports the use of multivariate machine‑learning models to capture complex interactions.

In [41]:
# sns.lmplot(data=df, x='Hardness', y='ph', hue='Potability', height=5, aspect=1.5)

# plt.show()

In [42]:
# sns.lmplot(data=df, x='Hardness', y='Chloramines', hue ='Potability', height=5, aspect=1.5)

# plt.show()

---

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.